lib/srv/heartbeatv2_test.go

```diff
@@ -0,0 +1,325 @@
+}
+
+func (h *fakeHeartbeatDriver) Announce(ctx context.Context, sender inventory.DownstreamSender) (ok bool) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	h.announceCount++
+	if h.announceErr > 0 {
+		h.announceErr--
+		return false
+	}
+	return true
+}
+
+func (h *fakeHeartbeatDriver) newStream(ctx context.Context, t *testing.T) client.UpstreamInventoryControlStream {
+	upstream, downstream := client.InventoryControlStreamPipe()
+	select {
+	case h.streamC <- downstream:
+	case <-ctx.Done():
+		t.Fatalf("context canceled during fake stream setup")
+	}
+
+	var msg proto.UpstreamInventoryMessage
+	select {
+	case msg = <-upstream.Recv():
+	case <-ctx.Done():
+		t.Fatalf("context canceled during fake stream recv")
+	}
+
+	_, ok := msg.(proto.UpstreamInventoryHello)
+	require.True(t, ok)
+
+	err := upstream.Send(ctx, proto.DownstreamInventoryHello{
+		ServerID: "test-auth",
+		Version:  teleport.Version,
+	})
+	require.NoError(t, err)
+
+	return upstream
+}
+
+func newFakeHeartbeatDriver() *fakeHeartbeatDriver {
+	// streamC is used to pass a fake control stream to the downstream handle's create func.
+	streamC := make(chan client.DownstreamInventoryControlStream)
+
+	hello := proto.UpstreamInventoryHello{
+		ServerID: "test-node",
+		Version:  teleport.Version,
+		Services: []types.SystemRole{types.RoleNode},
+	}
+
+	handle := inventory.NewDownstreamHandle(func(ctx context.Context) (client.DownstreamInventoryControlStream, error) {
+		// we're emulating an inventory.DownstreamCreateFunc here, but those are typically
+		// expected to return an error if no stream can be acquired. we're deliberately going
+		// with a blocking strategy instead here to avoid dealing w/ backoff that could make the
+		// test need to run longer.
+		select {
+		case <-ctx.Done():
+			return nil, trace.Errorf("context canceled while waiting for next control stream")
+		case stream := <-streamC:
+			return stream, nil
+		}
+	}, hello)
+
+	return &fakeHeartbeatDriver{
+		handle:      handle,
+		streamC:     streamC,
+		pollChanged: 1, // first poll is always true
+	}
+}
+
+// TestHeartbeatV2Basics verifies the basic functionality of HeartbeatV2 under various conditions by
+// injecting a fake driver and attempting to force the HeartbeatV2 into all of its happy and unhappy
+// states while monitoring test events to verify expected behaviors.
+func TestHeartbeatV2Basics(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// set up fake hb driver that lets us easily inject failures for
+	// the diff steps and assists w/ faking inventory control handles.
+	driver := newFakeHeartbeatDriver()
+	defer driver.handle.Close() // driver includes fake handle that needs cleanup
+
+	hb := newHeartbeatV2(driver.handle, driver, heartbeatV2Config{
+		announceInterval: time.Millisecond * 200,
+		pollInterval:     time.Millisecond * 50,
+		fallbackBackoff:  time.Millisecond * 400,
+		testEvents:       make(chan hbv2TestEvent, 1028),
+	})
+	go hb.Run()
```

Comment:
Any reason we can't have `hb.Run()` take a context? It would be nice to standardize and use context for the cancelation of any long-running operation.

Reply:
See https://github.com/gravitational/teleport/pull/12715#discussion_r893090454 for my thoughts on this.  Tl;dr is that I'm leaning toward the position that explicit `Close` is a better pattern unless there is a specific reason why the component in question is more suited to implicit closure.  Open to the idea that I'm way off base on that though.  Tell me what you think.